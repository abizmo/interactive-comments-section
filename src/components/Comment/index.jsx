import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { ReactComponent as Delete } from '../../assets/icons/delete.svg';
import { ReactComponent as Edit } from '../../assets/icons/edit.svg';
import { ReactComponent as Reply } from '../../assets/icons/reply.svg';
import Avatar from '../Avatar';
import Button from '../Button';
import EditComment from '../EditComment';
import NewComment from '../NewComment';
import Voting from '../Voting';
import {
  At, Label, Nickname, Wrapper,
} from './style';

function Comment({
  comment, currentUser, onDelete, onEdit, onReply, replyingTo, user,
}) {
  const [replying, setReplying] = useState(false);
  const [updating, setUpdating] = useState(false);
  const [votes, setVotes] = useState(comment.score || 0);

  const handleEdit = (c) => {
    onEdit(c);
    setUpdating(false);
  };

  const handleReply = (r) => {
    onReply(r, comment.id, user.username);
    setReplying(false);
  };

  const startEdit = () => {
    setUpdating(true);
  };

  const startReply = () => {
    setReplying(true);
  };

  const at = replyingTo && `@${replyingTo} `;
  const you = user.username === currentUser.username;

  return (
    <>
      <Wrapper>
        <header>
          <Avatar user={user} />
          <Nickname>{user.username}</Nickname>
          { you && (
          <Label>you</Label>
          )}
          <span>{comment.createdAt}</span>
        </header>
        <div className="content">
          {
          updating ? (
            <EditComment
              at={at}
              content={`${comment.content}`}
              onEdit={handleEdit}
            />
          ) : (
            <p>
              { replyingTo && (
                <At>
                  {at}
                </At>
              )}
              {comment.content}
            </p>
          )
        }
        </div>
        <div className="votes">
          <Voting onVote={setVotes} votes={votes} />
        </div>
        <div className="buttons">
          {
          you ? (
            <>
              <Button
                color="secondary"
                icon={Delete}
                label="Delete"
                onClick={() => onDelete(comment.id)}
              />
              <Button
                color="primary"
                icon={Edit}
                label="Edit"
                onClick={startEdit}
              />
            </>
          ) : (
            <Button
              color="primary"
              icon={Reply}
              label="Reply"
              onClick={startReply}
            />
          )
        }
        </div>
      </Wrapper>
      { replying && (
        <NewComment
          onCreate={handleReply}
          replyingTo={user.username}
          user={currentUser}
        />
      )}
    </>
  );
}

Comment.propTypes = {
  comment: PropTypes.shape({
    id: PropTypes.number.isRequired,
    content: PropTypes.string.isRequired,
    createdAt: PropTypes.string.isRequired,
    score: PropTypes.number,
  }).isRequired,
  currentUser: PropTypes.shape({
    image: PropTypes.shape().isRequired,
    username: PropTypes.string.isRequired,
  }).isRequired,
  onDelete: PropTypes.func.isRequired,
  onEdit: PropTypes.func.isRequired,
  onReply: PropTypes.func.isRequired,
  replyingTo: PropTypes.string,
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
  }).isRequired,
};

Comment.defaultProps = {
  replyingTo: '',
};

export default Comment;
