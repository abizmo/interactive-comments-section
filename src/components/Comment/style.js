import styled from 'styled-components';

export const Wrapper = styled.article`
  background: ${({ theme }) => theme.colors['neutral-100']};
  border-radius: .5rem;
  display: grid;
  gap: 1rem;
  grid-template-areas: 
    'header header'
    'content content'
    'votes buttons';
  grid-template-columns: repeat(2, 1fr);
  padding: 1rem;

  @media(${({ theme }) => theme.mediaQueries.lg}) {
    column-gap: 1.5rem;
    grid-template-areas:
      'votes header buttons'
      'votes content content';
    grid-template-columns: min-content 3fr 1fr;
    padding: 1.5rem;
  }

  & > header {
    align-items: center;
    grid-area: header;
    display: flex;
    gap: 1rem;
    line-height: 1.1;
  }

  & > #content {
    grid-area: content;
  }

  & > #votes {
    grid-area: votes;
  }

  & > #buttons {
    align-self: center;
    grid-area: buttons;
    justify-self: end;
    display: flex;
    gap: 1rem;
  }
`;

export const At = styled.span`
  color: ${({ theme }) => theme.colors['primary-500']};
  font-weight: 500;
`;

export const Nickname = styled.span`
  color: ${({ theme }) => theme.colors['neutral-800']};
  font-weight: 500;
  text-transform: lowercase;
`;

export const Label = styled.span`
  background: ${({ theme }) => theme.colors['primary-500']};
  border-radius: .25rem;
  color: ${({ theme }) => theme.colors['neutral-100']};
  font-size: ${({ theme }) => theme.fontSizes[300]};
  font-weight: 500;
  line-height: 1.1;
  padding: .2rem .5rem;
  text-align: center;
`;
