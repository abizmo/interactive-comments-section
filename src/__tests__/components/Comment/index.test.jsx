import React from 'react';
import {
  fireEvent, render, screen, cleanup,
} from '@testing-library/react';

import Comment from '../../../components/Comment';
import Theme from '../../../theme';

const comment = {
  id: 1,
  author: 'amyrobson',
  date: '1 month ago',
  body: 'Impressive!',
  votes: 0,
  replyingTo: 'maxblagun',
};

afterEach(cleanup);

describe('Comment from other user', () => {
  let handleDelete;
  let handleEdit;
  let handleReply;
  let rendered;

  beforeEach(() => {
    handleDelete = jest.fn();
    handleEdit = jest.fn();
    handleReply = jest.fn();

    rendered = render(
      <Theme>
        <Comment
          author={comment.author}
          date={comment.date}
          body={comment.body}
          likes={comment.votes}
          onDelete={handleDelete}
          onEdit={handleEdit}
          onReply={handleReply}
        />
      </Theme>,
    );
  });

  test('should render a comment', () => {
    const { container } = rendered;

    expect(screen.getByText(comment.author)).toBeTruthy();
    expect(screen.getByText(comment.body)).toBeTruthy();

    expect(screen.queryByText('Delete')).toBeFalsy();
    expect(screen.queryByText('you')).toBeFalsy();
    expect(screen.queryByText('@')).toBeFalsy();

    expect(container).toMatchSnapshot();
  });

  test('click reply button', () => {
    const button = screen.getByText('Reply');

    fireEvent.click(button);
    expect(handleReply).toHaveBeenCalled();
  });
});

describe('Comment from current user', () => {
  let handleDelete;
  let handleEdit;
  let handleReply;
  let rendered;

  beforeEach(() => {
    handleDelete = jest.fn();
    handleEdit = jest.fn();
    handleReply = jest.fn();

    rendered = render(
      <Theme>
        <Comment
          author={comment.author}
          date={comment.date}
          body={comment.body}
          likes={comment.votes}
          onDelete={handleDelete}
          onEdit={handleEdit}
          onReply={handleReply}
          you
        />
      </Theme>,
    );
  });

  test('should render a comment', () => {
    const { container } = rendered;

    expect(screen.queryByText('Reply')).toBeFalsy();
    expect(screen.queryByText('you')).toBeTruthy();

    expect(container).toMatchSnapshot();
  });

  test('click delete button', () => {
    const button = screen.getByText('Delete');

    fireEvent.click(button);
    expect(handleDelete).toHaveBeenCalled();
  });

  test('click edit button', () => {
    const button = screen.getByText('Edit');

    fireEvent.click(button);
    expect(handleEdit).toHaveBeenCalled();
  });
});

describe('Replied to some user', () => {
  let handleDelete;
  let handleEdit;
  let handleReply;

  beforeEach(() => {
    handleDelete = jest.fn();
    handleEdit = jest.fn();
    handleReply = jest.fn();

    render(
      <Theme>
        <Comment
          author={comment.author}
          date={comment.date}
          body={comment.body}
          likes={comment.votes}
          onDelete={handleDelete}
          onEdit={handleEdit}
          onReply={handleReply}
          replyingTo={comment.replyingTo}
        />
      </Theme>,
    );
  });

  test('should show the user who has been replied', () => {
    expect(screen.getByText(`@${comment.replyingTo}`)).toHaveStyle('color: hsl(238,40%, 52%)');
  });
});
