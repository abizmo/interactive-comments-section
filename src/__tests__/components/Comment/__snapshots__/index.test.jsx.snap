// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment from current user should render a comment 1`] = `
<div>
  <article
    class="sc-papXJ XSksx"
  >
    <header>
      <picture
        class="sc-bczRLJ kirzjg"
      >
        <source
          srcset="/./images/avatars/image-amyrobson.webp"
          type="image/webp"
        />
        <img
          alt="amyrobson"
          src="/./images/avatars/image-amyrobson.png"
          style="width: 100%;"
        />
      </picture>
      <h2
        class="sc-kDDrLX ikYFNm"
      >
        amyrobson
      </h2>
      <span
        class="sc-iqcoie ibqFpS"
      >
        you
      </span>
      <span>
        1 month ago
      </span>
    </header>
    <div
      class="content"
    >
      <p>
        Impressive! Though it seems the drag feature could be improved. But overall it looks incredible. You've nailed the design and the responsiveness at various breakpoints works really well.
      </p>
    </div>
    <div
      class="votes"
    >
      <div
        class="sc-iBkjds fTXLna"
        data-testid="voting"
      >
        <button
          class="sc-ftvSup jPGLgn"
          type="button"
        >
          <svg>
            plus.svg
          </svg>
          <span
            class="sr-only"
          >
            Plus
          </span>
        </button>
        <p>
          12
        </p>
        <button
          class="sc-ftvSup jPGLgn"
          type="button"
        >
          <svg>
            minus.svg
          </svg>
          <span
            class="sr-only"
          >
            Minus
          </span>
        </button>
      </div>
    </div>
    <div
      class="buttons"
    >
      <button
        class="sc-gsnTZi vpzmz"
        type="button"
      >
        <svg>
          delete.svg
        </svg>
        Delete
      </button>
      <button
        class="sc-gsnTZi dMLRiJ"
        type="button"
      >
        <svg>
          edit.svg
        </svg>
        Edit
      </button>
    </div>
  </article>
</div>
`;

exports[`Comment from other user should render a comment 1`] = `
<div>
  <article
    class="sc-papXJ XSksx"
  >
    <header>
      <picture
        class="sc-bczRLJ kirzjg"
      >
        <source
          srcset="/./images/avatars/image-amyrobson.webp"
          type="image/webp"
        />
        <img
          alt="amyrobson"
          src="/./images/avatars/image-amyrobson.png"
          style="width: 100%;"
        />
      </picture>
      <h2
        class="sc-kDDrLX ikYFNm"
      >
        amyrobson
      </h2>
      <span>
        1 month ago
      </span>
    </header>
    <div
      class="content"
    >
      <p>
        Impressive! Though it seems the drag feature could be improved. But overall it looks incredible. You've nailed the design and the responsiveness at various breakpoints works really well.
      </p>
    </div>
    <div
      class="votes"
    >
      <div
        class="sc-iBkjds fTXLna"
        data-testid="voting"
      >
        <button
          class="sc-ftvSup jPGLgn"
          type="button"
        >
          <svg>
            plus.svg
          </svg>
          <span
            class="sr-only"
          >
            Plus
          </span>
        </button>
        <p>
          12
        </p>
        <button
          class="sc-ftvSup jPGLgn"
          type="button"
        >
          <svg>
            minus.svg
          </svg>
          <span
            class="sr-only"
          >
            Minus
          </span>
        </button>
      </div>
    </div>
    <div
      class="buttons"
    >
      <button
        class="sc-gsnTZi dMLRiJ"
        type="button"
      >
        <svg>
          reply.svg
        </svg>
        Reply
      </button>
    </div>
  </article>
</div>
`;
